client/src/components/ExpandingFormset/index.js

```diff
@@ -0,0 +1,43 @@
+import $ from 'jquery';
+
+export class ExpandingFormset {
+  constructor(prefix, opts = {}) {
+    this.opts = opts;
+    const addButton = $('#' + prefix + '-ADD');
+    this.formContainer = $('#' + prefix + '-FORMS');
+    this.totalFormsInput = $('#' + prefix + '-TOTAL_FORMS');
+    this.formCount = parseInt(this.totalFormsInput.val(), 10);
+
+    if (opts.onInit) {
+      for (let i = 0; i < this.formCount; i += 1) {
+        opts.onInit(i);
+      }
+    }
+
+    const emptyFormElement = document.getElementById(
+      prefix + '-EMPTY_FORM_TEMPLATE',
+    );
+    if (emptyFormElement.innerText) {
+      this.emptyFormTemplate = emptyFormElement.innerText;
+    } else if (emptyFormElement.textContent) {
+      this.emptyFormTemplate = emptyFormElement.textContent;
+    }
+
+    addButton.on('click', () => {
+      this.addForm();
+    });
+  }
+
+  addForm() {
+    const newFormHtml = this.emptyFormTemplate
+      .replace(/__prefix__(.*?['"])/g, this.formCount + '$1')
+      .replace(/<-(-*)\/script>/g, '<$1/script>');
+
+    this.formContainer.append(newFormHtml);
+    if (this.opts.onAdd) this.opts.onAdd(this.formCount);
+    if (this.opts.onInit) this.opts.onInit(this.formCount);
```

Comment:
If we added an event dispatch here - you may get all you need to hook in the custom on add behaviour. Without needing to expose another global instance. 


```suggestion
    if (this.opts.onInit) this.opts.onInit(this.formCount);
    this.formContainer.dispatchEvent(new CustomEvent('wagtail:formset-add', {detail{formCount: this.formCount, formset:this,}, bubbles: true}));
```

Rough code only but this means that the custom multiple panel component just needs to add an event listener for this custom event and can get 'access' to this instance. 

See issue for notes about this approach (raised as InlinePanel but probably makes more sense in this code). https://github.com/wagtail/wagtail/issues/9105

Also - this likely is where the code will head with Stimulus - DOM event driven, to support native behaviour and less globals + minimal exposure of internal JS classes where possible. 

You can even use a cancellabe event to stop default behaviour. 

Not suggesting this event is added in this PR but just flagging while this code is being looked at.